File: 2026POTD/JANUARY/09-01-2026/865.ex
defmodule Solution do
  @spec subtree_with_all_deepest(root :: TreeNode.t | nil) :: TreeNode.t | nil
  def subtree_with_all_deepest(root) do
    dfs = fn dfs, node ->
      if node == nil do
        {0, nil}
      else
        {ld, ln} = dfs.(dfs, node.left)
        {rd, rn} = dfs.(dfs, node.right)
        cond do
          ld > rd -> {ld + 1, ln}
          rd > ld -> {rd + 1, rn}
          true -> {ld + 1, node}
        end
      end
    end

    {_, result} = dfs.(dfs, root)
    result
  end
end

File: 2026POTD/JANUARY/09-01-2026/865.go
func subtreeWithAllDeepest(root *TreeNode) *TreeNode {
    var dfs func(*TreeNode) (int, *TreeNode)
    dfs = func(node *TreeNode) (int, *TreeNode) {
        if node == nil {
            return 0, nil
        }
        ld, ln := dfs(node.Left)
        rd, rn := dfs(node.Right)
        if ld > rd {
            return ld + 1, ln
        }
        if rd > ld {
            return rd + 1, rn
        }
        return ld + 1, node
    }
    _, res := dfs(root)
    return res
}